compiler/noirc_evaluator/src/ssa/acir_gen/mod.rs

```diff
@@ -976,16 +976,39 @@ impl<'a> Context<'a> {
             }
         };
 
-        if self.handle_constant_index(instruction, dfg, index, array, store_value)? {
-            return Ok(());
+        let array_id = dfg.resolve(array);
+        let array_typ = dfg.type_of_value(array_id);
+        // Compiler sanity checks
+        assert!(!array_typ.is_nested_slice(), "ICE: Nested slice type has reached ACIR generation");
+        let (Type::Array(_, _) | Type::Slice(_)) = &array_typ else {
+            unreachable!("ICE: expected array or slice type");
+        };
+
+        match self.convert_value(array_id, dfg) {
+            AcirValue::Var(acir_var, _) => {
+                return Err(RuntimeError::InternalError(InternalError::Unexpected {
+                    expected: "an array value".to_string(),
+                    found: format!("{acir_var:?}"),
+                    call_stack: self.acir_context.get_call_stack(),
+                }))
+            }
+            AcirValue::Array(array) => {
+                // `AcirValue::Array` supports reading/writing to constant indices at compile-time in some cases.
+                if let Some(constant_index) = dfg.get_numeric_constant(index) {
+                    let store_value = store_value.map(|value| self.convert_value(value, dfg));
+                    if self.handle_constant_index(instruction, dfg, array, constant_index, store_value)? {
+                         return Ok(());
+                    }
+                }
+            }
+            AcirValue::DynamicArray(_) => (),
```

Comment:
This function is getting large - can you move this parsing code out of it?
We could have a helper for `try_handle_constant_index` or similar returning a `Result<Option<acir array>, error>` maybe.

Reply:
`handle_constant_index` already returns a result. We could just move these checks back into there.

Reply:
Hmm, yeah it looks like there could be some repeated code in this match in handle_constant_index as well. If it's cleaner it can go there, otherwise another helper would be good too. Whichever approach is cleaner.

Reply:
I had looked at this but I don't like how we'd be returning an `AcirValue` which could potentially be either an updated array or a single value read from that array. Realistically we need to separate handling for reads from writes more cleanly and that would be OOS for this PR.

Reply:
I guess we can just move this to a helper then for now, and maybe comment that we want to split reads/writes more cleanly in the future.